core/src/main/scala/org/apache/spark/sql/execution/columnar/encoding/Uncompressed.scala

```diff
@@ -34,256 +33,165 @@ trait Uncompressed extends ColumnEncoding {
-
-  override def readLong(columnBytes: AnyRef, cursor: Long): Long =
-    ColumnEncoding.readLong(columnBytes, cursor)
+  override def readShort(columnBytes: AnyRef, nonNullPosition: Int): Short =
+    ColumnEncoding.readShort(columnBytes, baseCursor + (nonNullPosition << 1))
 
-  override def nextFloat(columnBytes: AnyRef, cursor: Long): Long =
-    cursor + 4
-
-  override def absoluteFloat(columnBytes: AnyRef, position: Int): Long =
-    baseCursor + ((position - numNullsUntilPosition(columnBytes, position)) << 2)
+  override def readInt(columnBytes: AnyRef, nonNullPosition: Int): Int =
+    ColumnEncoding.readInt(columnBytes, baseCursor + (nonNullPosition << 2))
 
-  override def readFloat(columnBytes: AnyRef, cursor: Long): Float =
-    ColumnEncoding.readFloat(columnBytes, cursor)
-
-  override def nextDouble(columnBytes: AnyRef, cursor: Long): Long =
-    cursor + 8
+  override def readLong(columnBytes: AnyRef, nonNullPosition: Int): Long =
+    ColumnEncoding.readLong(columnBytes, baseCursor + (nonNullPosition << 3))
 
-  override def absoluteDouble(columnBytes: AnyRef, position: Int): Long =
-    baseCursor + ((position - numNullsUntilPosition(columnBytes, position)) << 3)
-
-  override def readDouble(columnBytes: AnyRef, cursor: Long): Double =
-    ColumnEncoding.readDouble(columnBytes, cursor)
-
-  override def nextLongDecimal(columnBytes: AnyRef, cursor: Long): Long =
-    cursor + 8
+  override def readFloat(columnBytes: AnyRef, nonNullPosition: Int): Float =
+    ColumnEncoding.readFloat(columnBytes, baseCursor + (nonNullPosition << 2))
 
-  override def absoluteLongDecimal(columnBytes: AnyRef, position: Int): Long =
-    baseCursor + ((position - numNullsUntilPosition(columnBytes, position)) << 3)
+  override def readDouble(columnBytes: AnyRef, nonNullPosition: Int): Double =
+    ColumnEncoding.readDouble(columnBytes, baseCursor + (nonNullPosition << 3))
 
   override def readLongDecimal(columnBytes: AnyRef, precision: Int,
-      scale: Int, cursor: Long): Decimal =
-    Decimal.createUnsafe(ColumnEncoding.readLong(columnBytes, cursor),
-      precision, scale)
-
-  override def nextDecimal(columnBytes: AnyRef, cursor: Long): Long = {
-    // cursor == 0 indicates first call so don't increment cursor
-    if (cursor != 0) {
-      cursor + 4 + ColumnEncoding.readInt(columnBytes, cursor)
-    } else {
-      baseCursor
+      scale: Int, nonNullPosition: Int): Decimal =
+    Decimal.createUnsafe(ColumnEncoding.readLong(columnBytes,
+      baseCursor + (nonNullPosition << 3)), precision, scale)
+
+  private def setCursorForVariableWidth(columnBytes: AnyRef, nonNullPosition: Int,
+      sizeWidth: Int = 4): Unit = {
+    // check sequential calls else skip as much required
+    var lastPosition = lastNonNullPosition
+    if (nonNullPosition != lastPosition + 1) {
+      var cursor = currentCursor
+      lastPosition += 1
+      do {
+        cursor += sizeWidth + ColumnEncoding.readInt(columnBytes, cursor)
```

Comment:
Good idea. Can even use varInt encoding like in protobuf/kryo which will use 1-5 bytes.